src/envs/behavior_options.py

```diff
@@ -0,0 +1,1012 @@
+import scipy
+from predicators.src.structs import State
+
+try:
+    import pybullet as p  # type: ignore
+    from igibson import object_states
+    from igibson.envs.behavior_env import BehaviorEnv
+    from igibson.external.pybullet_tools.utils import CIRCULAR_LIMITS
+    from igibson.objects.articulated_object import URDFObject
+    from igibson.utils.behavior_robot_planning_utils import (
+        plan_base_motion_br,
+        plan_hand_motion_br,
+    )
+    from igibson.utils import sampling_utils
+    from igibson.external.pybullet_tools.utils import (
+        get_aabb,
+        get_aabb_extent,
+    )
+except ModuleNotFoundError as e:
+    print(e)
+
+_ON_TOP_RAY_CASTING_SAMPLING_PARAMS = {
+    "max_angle_with_z_axis": 0.17,
+    "bimodal_stdev_fraction": 1e-6,
+    "bimodal_mean_fraction": 1.0,
+    "max_sampling_attempts": 50,
+    "aabb_offset": 0.01,
+}
+
+
+def get_body_ids(  # type: ignore
+    env,
+    include_self: bool = False,
+    grasping_with_right: bool = False,
+) -> List[int]:
+    """Function to return list of body_ids for all objects for collision
+    checking depending on whether navigation or grasping/placing is being
+    done."""
+    ids = []
+    for obj in env.scene.get_objects():
+        if isinstance(obj, URDFObject):
+            # We want to exclude the floor since we're always floating and
+            # will never practically collide with it, but if we include it
+            # in collision checking, we always seem to collide.
+            if obj.name != "floors":
+                ids.extend(obj.body_ids)
+
+    if include_self:
+        ids.append(env.robots[0].parts["left_hand"].get_body_id())
+        ids.append(env.robots[0].parts["body"].get_body_id())
+        ids.append(env.robots[0].parts["eye"].get_body_id())
+        if not grasping_with_right:
+            ids.append(env.robots[0].parts["right_hand"].get_body_id())
+
+    return ids
+
+
+def detect_collision(bodyA: int, object_in_hand: int = None) -> bool:
+    """Detects collisions between objects in the scene (except for the object
+    in the robot's hand)"""
+    collision = False
+    for body_id in range(p.getNumBodies()):
+        if body_id in [bodyA, object_in_hand]:
+            continue
+        closest_points = p.getClosestPoints(bodyA, body_id, distance=0.01)
+        if len(closest_points) > 0:
+            collision = True
+            break
+    return collision
+
+
+def detect_robot_collision(robot) -> bool:  # type: ignore
+    """Function to detect whether the robot is currently colliding with any
+    object in the scene."""
+    object_in_hand = robot.parts["right_hand"].object_in_hand
+    return (detect_collision(robot.parts["body"].body_id)
+            or detect_collision(robot.parts["left_hand"].body_id)
+            or detect_collision(robot.parts["right_hand"].body_id,
+                                object_in_hand))
+
+
+def get_aabb_volume(lo: np.ndarray, hi: np.ndarray) -> float:
```

Comment:
instead of `np.ndarray` for consistency can you use the type `Array`? just `git grep "Array"` to see what I mean, it's defined near the bottom of structs.py

Reply:
Looks like `np.ndarray` is used a lot in this file, maybe just find-and-replace?